src/systems/inter_mesh_projection.C

```diff
@@ -38,27 +49,144 @@ namespace libMesh
+
+        // Construct local version of the current system vector
+	    // This has to be a serial vector
+        std::unique_ptr<NumericVector<Number>> solution_vector_serial = NumericVector<Number>::build(from_system.comm());
+	    solution_vector_serial->init(from_system.solution->size(), true, SERIAL);
+
+        std::vector<Number> solution_vector;
+        from_system.update_global_solution(solution_vector);
+        (*solution_vector_serial) = solution_vector;
+
+        // Construct a MeshFunction for the solution
+        MeshFunction * mesh_func_solution =
+            new MeshFunction(from_system.get_equation_systems(), *solution_vector_serial,
+                             from_system.get_dof_map(), variables_vector);
+
+        mesh_func_solution->init();
+
+        // For some element types (say C1) we also need to pass a gradient evaluation MeshFunction
+        // To do this evaluate, a new shim class GradientMeshFunction has been added which redirects
+        // gptr::operator evaluations inside projection methods into MeshFunction::gradient calls.
+        GradientFunction * gptr_solution = new GradientFunction(mesh_func_solution);
+        gptr_solution->init();
+
+        from_system.update();
+
+        // // Testing
+        // Point p_eval_sol;
+        // p_eval_sol(0) = 0.6;
+        // //p_eval(1) = 0.5;
+
+        // DenseVector<Number> vector_point_value_sol;
+        // mesh_func_solution->operator()(p_eval_sol, 0.0, vector_point_value_sol);
+
+        // std::cout<<"vector_point_value_sol(0.6): "<<vector_point_value_sol(0)<<std::endl;
+
+        // DenseVector<Gradient> vector_gradient_from_mesh_func_sol;
+        // mesh_func_solution->gradient(p_eval_sol, 0.0, vector_gradient_from_mesh_func_sol.get_values());
+        // std::cout<<"gradient_point_value_from_mesh_func_sol(0.6): "<<vector_gradient_from_mesh_func_sol(0)(0)<<std::endl;
+
+        // DenseVector<libMesh::Gradient> vector_gradient_from_gptr_sol;
+        // gptr_solution->operator()(p_eval_sol, 0.0, vector_gradient_from_gptr_sol);
+        // std::cout<<"gradient_point_value_sol(0.6): "<<vector_gradient_from_gptr_sol(0)(0)<<std::endl;
+
+        Real T_from_H1 = from_system.calculate_norm(*from_system.solution, 0, H1);
+        Real T_from_L2 = from_system.calculate_norm(*from_system.solution, 0, L2);
+
+        // Print the vector
+        for(unsigned int i = 0; i < from_system.solution->size(); i++)
+            std::cout<<(*from_system.solution)(i)<<std::endl;
+
+        std::cout << "T_from(H1) = " << T_from_H1 << std::endl << "T_from(L2) = " << T_from_L2 << std::endl;
+
+        // End testing
+
+        // FIXME: To properly apply adjoint Dirichlet constraints, we need to pass a fourth argument when we are projecting adjoint vectors.
+        to_system.project_vector(*to_system.solution, dynamic_cast<FunctionBase<Number> *>(mesh_func_solution), dynamic_cast<FunctionBase<Gradient> *>(gptr_solution));
```

Comment:
Why do you have dynamic_cast here?  C++ doesn't need any sort of casting to interpret a subclass pointer as a base class pointer.

Reply:
Fixed in upcoming commit.